Sources/PackageCollectionsSigning/Certificate/CertificatePolicy.swift

```diff
@@ -82,19 +95,329 @@ extension CertificatePolicy {
+                     httpClient: HTTPClient,
+                     diagnosticsEngine: DiagnosticsEngine,
+                     callbackQueue: DispatchQueue,
+                     callback: @escaping (Result<Void, Error>) -> Void) {
+        let wrappedCallback: (Result<Void, Error>) -> Void = { result in callbackQueue.async { callback(result) } }
+
+        let ocspURLs = certificate.withUnsafeMutablePointer { CCryptoBoringSSL_X509_get1_ocsp($0) }
+        defer { CCryptoBoringSSL_sk_OPENSSL_STRING_free(ocspURLs) }
+
+        let ocspURLCount = CCryptoBoringSSL_sk_OPENSSL_STRING_num(ocspURLs)
+        // Nothing to do if no OCSP URLs. Use `supportsOCSP` to require OCSP support if needed.
+        guard ocspURLCount > 0 else { return wrappedCallback(.success(())) }
+
+        // Construct the OCSP request
+        let digest = CCryptoBoringSSL_EVP_sha1()
+        let certid = certificate.withUnsafeMutablePointer { certPtr in
+            issuer.withUnsafeMutablePointer { issPtr in
+                OCSP_cert_to_id(digest, certPtr, issPtr)
+            }
+        }
+        let request = OCSP_REQUEST_new()
+        defer { OCSP_REQUEST_free(request) }
+
+        guard OCSP_request_add0_id(request, certid) != nil else {
+            return wrappedCallback(.failure(CertificatePolicyError.ocspSetupFailure))
+        }
+
+        // Write the request binary to memory bio
+        let bio = CCryptoBoringSSL_BIO_new(CCryptoBoringSSL_BIO_s_mem())
+        defer { CCryptoBoringSSL_BIO_free(bio) }
+        guard i2d_OCSP_REQUEST_bio(bio, request) > 0 else {
+            return wrappedCallback(.failure(CertificatePolicyError.ocspSetupFailure))
+        }
+
+        // Copy from bio to byte array then convert to Data
+        var count = 0
+        var out: UnsafePointer<UInt8>?
+        guard CCryptoBoringSSL_BIO_mem_contents(bio, &out, &count) > 0 else {
+            return wrappedCallback(.failure(CertificatePolicyError.ocspSetupFailure))
+        }
+
+        let requestData = Data(UnsafeBufferPointer(start: out, count: count))
+
+        let results = ThreadSafeArrayStore<Result<Bool, Error>>()
+        let group = DispatchGroup()
+
+        // Query each OCSP responder and record result
+        for index in 0 ..< ocspURLCount {
+            guard let urlStr = CCryptoBoringSSL_sk_OPENSSL_STRING_value(ocspURLs, numericCast(index)),
+                let url = String(validatingUTF8: urlStr).flatMap({ URL(string: $0) }) else {
+                results.append(.failure(OCSPError.badURL))
+                continue
+            }
+
+            let cacheKey = CacheKey(url: url, request: requestData)
+            if let cachedResult = self.resultCache[cacheKey] {
+                if cachedResult.timestamp + self.cacheTTL > DispatchTime.now() {
+                    results.append(.success(cachedResult.isCertGood))
+                    continue
+                } else {
+                    _ = self.resultCache.removeValue(forKey: cacheKey)
```

Comment:
Technically speaking there's a time-of-check-time-of-use problem here where we could race this removal with an insertion from a different thread. Do we care?

Reply:
The remove is not really necessary so I am deleting it.